---
title: "Exploratory Data Analysis (EDA) on bulk RNA-seq from in-vitro Myofibroblast with different perturbations on NKD2"
author: "Javier Perales-Pat√≥n - javier.perales@bioquant.uni-heidelberg.de - ORCID: 0000-0003-0780-6683" 
output: github_document
---

Transcriptome profiling with conventional RNAseq was performed in three replicates of myofibroblasts 
carrying different gene perturbationss on NKD2 (Knock-out, over-expression and corresponding controls with 
empty vector). These belong to two different experiments: knock-out (KO) and over-dispersion (OE). Thus, these 
are loaded at once, but explored and analyzed independently. In the case of KO experiment, there were generated 
3 NKD2-KO clones, one of them (`nC3`) with a phenotype with a severe collagen production.
Herein we perform an exploratory analysis. 

## Setting-up environment
The environment will be set with a random seed number for reproducibility and an output folder
for processed data and figures.

### set env
```{r env}
options(stringsAsFactors = FALSE)
# Seed number
set.seed(1234)
# Output directory
OUTDIR <- "./00_EDA_output/"
if(!dir.exists(OUTDIR)) dir.create(OUTDIR);

# Figures
FIGDIR <- paste0(OUTDIR, "/figures/")
knitr::opts_chunk$set(fig.path=FIGDIR)
knitr::opts_chunk$set(dev=c('png', 'pdf'))
# Data
DATADIR <- paste0(OUTDIR, "/data/")
if(!dir.exists(DATADIR)) dir.create(DATADIR);

# If already exists, clean dirs?
clean_dirs <- TRUE
if(clean_dirs) {
	unlink(list.files(OUTDIR, full.names=TRUE, recursive = TRUE))
}
# Color palette
cols <- c("grey","#cb6a49", "#a46cb7","#7aa457")
```

### Load libraries
Essential libraries for R analysis.
```{r}
library(edgeR)
library(limma)
library(fgsea)
library(GSEABase)
library(ggplot2)
library(cowplot)
library(SummarizedExperiment)
library(csaw)
fontTXT <- "sans"
```

## Load data and normalize

```{r}
### 1 Load data
# Source: https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#limma-voom
load("../data/bulk-rnaseq/forJavier.RData")
# The sample IDs are stored in the folder from the file that was imported to R. We recall them
sname <- sapply(dirname(filenames),function(z) paste(strsplit(z, split="_")[[1]][1:2],collapse="_"))
# We get the read counts per sample
cnt <- txi$counts
# Rename samples
colnames(cnt) <- sname
# Get the biological conditions
group_str <- as.character(group)
group_str <- gsub("-","_", group_str)
group_str <- gsub("OverExpr","OE",group_str)
```

### Knock-out experiment
Define data experiment
```{r}
KO_idx <- grep("^KO", group_str)
KO_cnt <- cnt[, KO_idx]
KO_gr <- factor(group_str[KO_idx])
KO_gr <- relevel(KO_gr, ref="KO_ctrl")
```

```{r}
## 2 Create DGElist (edgeR package, followed by TMM default normalization method)
KO_y <- DGEList(counts = KO_cnt, group = KO_gr, genes = rownames(KO_cnt))
# Filtering
KO_keep <- filterByExpr(KO_y)
table(KO_keep)

# Apply filtering
KO_y <- KO_y[KO_keep, ]
KO_y <- calcNormFactors(KO_y)

## 3 Create experimental design
stopifnot(all(KO_gr == KO_y$samples$group))
KO_design <- model.matrix(~ 0 + KO_gr) # wo/ intercept
colnames(KO_design) <- gsub("^KO_gr", "", colnames(KO_design))
KO_v <- voom(KO_y, KO_design)
```

### Over-expression experiment
Define data experiment
```{r}
OE_idx <- grep("^OE", group_str)
OE_cnt <- cnt[, OE_idx]
OE_gr <- factor(group_str[OE_idx])
OE_gr <- relevel(OE_gr, ref="OE_ctrl")
```

```{r}
## 2 Create DGElist (edgeR package, followed by TMM default normalization method)
OE_y <- DGEList(counts = OE_cnt, group = OE_gr, genes = rownames(OE_cnt))
# Filtering
OE_keep <- filterByExpr(OE_y)
table(OE_keep)

# Apply filtering
OE_y <- OE_y[OE_keep, ]
OE_y <- calcNormFactors(OE_y)

## 3 Create experimental design
stopifnot(all(OE_gr == OE_y$samples$group))
OE_design <- model.matrix(~ 0 + OE_gr) # wo/ intercept
colnames(OE_design) <- gsub("^OE_gr", "", colnames(OE_design))
OE_v <- voom(OE_y, OE_design)
```

## Diagnostics
### The library size
```{r}
par(mar=c(10,8,4,4)); 
barplot(c(KO_y$samples$lib.size, OE_y$samples$lib.size),
	names.arg = c(rownames(KO_y$samples), rownames(OE_y$samples)), 
	las=2)
```

### NKD2 expression

```{r NKD2_expr, fig.width=6, fig.height=4, dpi=300}
# Source: https://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#edgeR
cts <- txi$counts
normMat <- txi$length

# Obtaining per-observation scaling factors for length, adjusted to avoid
# changing the magnitude of the counts.
normMat <- normMat/exp(rowMeans(log(normMat)))
normCts <- cts/normMat

# Computing effective library sizes from scaled counts, to account for
# composition biases between samples.
eff.lib <- calcNormFactors(normCts) * colSums(normCts)

# Combining effective library sizes with the length factors, and calculating
# offsets for a log-link GLM.
normMat <- sweep(normMat, 2, eff.lib, "*")
normMat <- log(normMat)

# Creating a DGEList object for use in edgeR.
y <- DGEList(cts)
y <- scaleOffset(y, normMat)

se <- SummarizedExperiment(assays = list(counts = y$counts, offset = y$offset))
se$totals <- y$samples$lib.size
cpms <- calculateCPM(se, use.offsets = TRUE, log = FALSE)
#

dat <- data.frame("NKD2"=cpms[grep("NKD2", rownames(cpms)),],
		  "Group"=group_str)
dat$Group <- factor(dat$Group, levels=c("KO_ctrl", "KO_1", "KO_2", "KO_3",
					"OE_ctrl", "OE")
)

ggplot(dat, aes(y=NKD2, x=Group, colour=Group)) + 
	geom_point(position=position_jitter(w=0.2, seed = 2354), size=2) +
	ggtitle("NKD2 quantification") +
	xlab("Condition") + ylab("NKD2 expression (CPM)") +
	theme_cowplot()
```


> Conclusion: Libraries sizes are well balanced. Surprisingly, NKD2 transcripts have 
not been sequenced in control conditions. Maybe the cell line presents a baseline 
protein levels of NKD2 to be funcional as it is shown in the western blot of these
samples.

### Principal Component analysis
First we start with the definition of ad-hoc handle functions to quick generate PCA plots 
(i.e. two-first principal components showing their variance explained).
```{r}
# Define some functions
runPCA <- function(mat) {
  pca <- prcomp(t(mat))
  
  return(pca)
}

plotPCA <- function(pca, dims, pchs, cols, labels=NULL, family="sans") {
  importance <- summary(pca)$importance[,dims]
  PCscores <- data.frame(pca$x)[,dims]
  
  plot(PCscores,las=1,
       pch=pchs,bg=cols,
       family=family,
       cex.axis=1.3, cex.lab=1.3,
       xlab=paste0(dims[1]," : ",format(round(importance[2,dims[1]]*100,digits = 2),nsmall = 2),"% var expl."),
       ylab=paste0(dims[2]," : ",format(round(importance[2,dims[2]]*100,digits = 2),nsmall = 2),"% var expl."))
  if(!is.null(labels)) {
    par(xpd=TRUE)
    text(x=PCscores[,dims[1]], y=PCscores[,dims[2]], labels=labels, pos=1, family=family)
  }
}

```

### Knock-out experiment
```{r pca_KO, fig.width=7, fig.height=5, dpi=300}
par(mar=c(4,4,4,12), xpd=TRUE)
# KO
plotPCA(pca=runPCA(mat = KO_v$E),
        dims=c("PC1","PC2"),
        pchs=c(21), cols=cols[as.integer(KO_gr)])
legend("topright",legend = levels(KO_gr), pch=c(21),
       pt.bg = cols[1:length(levels(KO_gr))], 
       inset = c(-0.4,0), title = "Perturbation")
```

```{r pca_OE, fig.width=7, fig.height=5, dpi=300}
par(mar=c(4,4,4,12), xpd=TRUE)
# OE
plotPCA(pca=runPCA(mat = OE_v$E),
        dims=c("PC1","PC2"),
        pchs=c(21), cols=cols[as.integer(OE_gr)])
legend("topright",legend = levels(OE_gr), pch=c(21),
       pt.bg = cols[1:length(levels(OE_gr))], 
       inset = c(-0.4,0), title = "Perturbation")
```

We conclude that biological replicates and conditions are cluster as it would be expected given 
the experimental settings. The NKD2-KO clone with more severe phenotype (`nC3`) is much more 
distinct to the other phenotypes. And KO and OE experiments cluster with their corresponding 
replicates.


## Session info

```{r}
sessionInfo()

{                                                                                                                                                                                                           
sink(file=paste0(OUTDIR,"/sessionInfo.txt"))
print(sessionInfo())
sink()
}
```

