---
title: "Inference of Transcription Factor activity in Myofibroblasts with NDK2 perturbations in kidney fibrosis"
author: "Javier Perales-Pat√≥n - javier.perales@bioquant.uni-heidelberg.de - ORCID: 0000-0003-0780-6683" 
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Herein we perform a Transcription Factor (TF) analysis on the role of NKD2 myofibroblasts during
kidney fibrosis. For that, we use normalized data from the differential gene expression analysis 
and estimate TF activities using DoRothEA mouse regulons.

## Setup environment
The environment will be set with a random seed number for reproducibility and an output folder
for processed data and figures.

### set env
```{r env}
options(stringsAsFactors = FALSE)
# Seed number
set.seed(1234)
# Output directory
OUTDIR <- "./03_TF_output/"
if(!dir.exists(OUTDIR)) dir.create(OUTDIR);

# Figures
FIGDIR <- paste0(OUTDIR, "/figures/")
knitr::opts_chunk$set(fig.path=FIGDIR)
knitr::opts_chunk$set(dev=c('png','pdf'))
# Data
DATADIR <- paste0(OUTDIR, "/data/")
if(!dir.exists(DATADIR)) dir.create(DATADIR);
# If already exists, clean dirs?
clean_dirs <- TRUE
if(clean_dirs) {
	unlink(list.files(OUTDIR, full.names=TRUE, recursive = TRUE))
}
```

### Load libraries
Essential libraries for R analysis.
```{r load_libs}
library(fgsea)
library(limma)
library(viper)
library(purrr)
library(dorothea)
library(ComplexHeatmap)
library(ggplot2)
library(cowplot)
library(ggrepel)
fontTXT <- "sans"
```

## Load data
We use the output data from DGE step (01). In particular,
the limma objects which contains gene expression, statistics and 
design of the contrasts.
```{r}
# Knock-out experiment
KO_v <- readRDS("./01_DGE_output/data/KO_v.rds")
KO_eBay.genes <- readRDS("./01_DGE_output/data/KO_eBay.rds")

# Over-expression experiment
OE_v <- readRDS("./01_DGE_output/data/OE_v.rds")
OE_eBay.genes <- readRDS("./01_DGE_output/data/OE_eBay.rds")
```

## Transcription factor analysis using DoRothEA
```{r dorothea}
# We read dorothea regulons for human
dorothea_regulon <- get(data("dorothea_hs", package="dorothea"))
# Obtain regulons based on interactions with confidence classes A,B,C
regul <- dorothea_regulon %>% dplyr::filter(confidence %in% c("A","B","C"))
# Compute VIPER scores

## Knock-out experiment
KO_TF <- run_viper(KO_v$E, regul,
		options = list(method="scale", minsize=4,
			       eset.filter=FALSE, cores=1,
			       verbose=FALSE))
## Over-expression experiment
OE_TF <- run_viper(OE_v$E, regul,
		options = list(method="scale", minsize=4,
			       eset.filter=FALSE, cores=1,
			       verbose=FALSE))

```

A handy limma test function
```{r limma_test}
eBayes_limma <- function(mat, design, cont) {
	fit <- lmFit(mat, design)
	fit2 <- contrasts.fit(fit, cont)
	eBay <- eBayes(fit2)
	return(eBay)
}
```


The TF activities are estimated using `viper`. Similarly to gene-level differential expression, 
contrasts performed using Empirical Bayes method implemented in `limma`
```{r}
## Knock-out experiment
KO_cont.mat <- makeContrasts("NKD2_KO_severe"=KO_1 - KO_ctrl,
			  "NKD2_KO_shallow"=(KO_2 + KO_3)/2 - KO_ctrl,
                          levels=KO_v$design)
KO_eBay <- eBayes_limma(KO_TF, KO_v$design, KO_cont.mat)

## Over-expression experiment
OE_cont.mat <- makeContrasts("NKD2_OE"=OE - OE_ctrl, levels=OE_v$design)
OE_eBay <- eBayes_limma(OE_TF, OE_v$design, OE_cont.mat)
```


## Broad statistics on diff TFs
Just check how many TFs are differentially active
```{r}
KO_diffTF_cnt <- apply(decideTests(KO_eBay),2, table)
OE_diffTF_cnt <- apply(decideTests(OE_eBay),2, table)
print(cbind(KO_diffTF_cnt, OE_diffTF_cnt))
```

## Diagnostics and save results
Write tables as supplementary data for the records and further interpretation.
```{r show_diffTF}
show_diffTF <- function(topTab, cont, DATADIR) {
  cat(paste0("Registering differential expression for ",cont,"\n"),
	file=stdout())
  # DEGs table
  write.table(topTab, file=paste0(DATADIR, cont,"_diffTF.csv"),sep=",",
                                  row.names=TRUE, col.names=NA, quote=FALSE)
  # Volcano plot
  plot(topTab$logFC, -log10(topTab$P.Value),
       xlab="log2-fold-change", ylab="-log10(pvalue)", cex=0.7)
  # Histogram of p-vals
  hist(topTab$P.Value)
}
```

```{r diffTF}
for(cont in colnames(KO_cont.mat)) {
	topTab <- topTable(KO_eBay, coef=cont, number = Inf)	
	show_diffTF(topTab, cont, DATADIR)
}
for(cont in colnames(OE_cont.mat)) {
	topTab <- topTable(OE_eBay, coef=cont, number = Inf)	
	show_diffTF(topTab, cont, DATADIR)
}
```
Histogram of p-values looks good, and volcano plots as well. Certainly Fibroblasts presents 
higher proportion of TF dysregulated from the medium conditioned with Macrophages PHD2cKO.

## Visualization

```{r volcano_TF, fig.width=14, fig.height=6, dpi=300}
dat <- list("KO_severe"=topTable(KO_eBay, coef="NKD2_KO_severe", number=Inf),
	    "KO_shallow"=topTable(KO_eBay, coef="NKD2_KO_shallow", number=Inf),
	    "OE"=topTable(OE_eBay, coef="NKD2_OE", number=Inf))
for(cnt in names(dat)) dat[[cnt]]$contrast <- cnt
for(cnt in names(dat)) dat[[cnt]]$ID <- rownames(dat[[cnt]])
for(cnt in names(dat)) {
	dat[[cnt]]$show <- FALSE
	dat[[cnt]]$show[1:20] <- TRUE
}

dat <- do.call("rbind", dat)
dat$signif <- ifelse(dat$adj.P.Val < 0.05, "FDR<0.05", "ns")

ggplot(dat, aes(x=logFC, y=-log10(P.Value), colour=signif)) +
	geom_point() +
	geom_label_repel(data=dat[dat$show & dat$logFC < 0,],
                          aes(label=ID), 
			  family=fontTXT, size=4,
                          force=2,
			  color="blue",
			  xlim=c(-10,0),
			  ylim=c(3,15),
			  segment.size = 0.5) +
	geom_label_repel(data=dat[dat$show & dat$logFC > 0,],
                          aes(label=ID), 
			  family=fontTXT, size=4,
                          force=1,
			  color="red",
			  xlim=c(0,10),
			  ylim=c(1,20)) +

    coord_cartesian(xlim = c(-13,10), 
		    clip = "off") +
    scale_color_manual(values=c("orange", "black")) +
    xlab("log2-fold-change") +
    theme_cowplot() + 
    theme(text = element_text(family=fontTXT, size=20),
	  legend.text = element_text(family=fontTXT, size=20),
	  legend.position = "bottom",
	  axis.text = element_text(family=fontTXT, size=20),
	  ) +
	facet_wrap(~ contrast)
```

## Downstream TFs of WNT pathway
NKD2 is a well-known master regulator of WNT pathway. It controls the WNT canonical pathway (inhibitor) 
and the Planar Cell Polarity pathway (activator). In this section we test for a differential activity of
TFs that are downstream of both pathways

### WNT Canonical pathway
TCF7 seems to be dysregulated in the opposite direction as it would be expected. NKD2 is a negative regulator of
upstream signal, and it would be expected that TCF7 is up-regulated if NKD2 is knock-out. But it shows the opposite.
LEF1 is up-regulated towards NKD2 over-expression, which is the opposite again.
```{r WNT_canonical}
# TCF7
dat[dat$ID=="TCF7",c(1,3,4,5,7,8)]
dat[dat$ID=="TCF7L1",c(1,3,4,5,7,8)]
dat[dat$ID=="TCF7L2",c(1,3,4,5,7,8)]
# LEF1
dat[dat$ID=="LEF1",c(1,3,4,5,7,8)]
```

Let's take a look at the leading edge of TCF7 targets:
```{r warning=FALSE}
TCF7_regul <- regul %>% dplyr::filter(tf == "TCF7")
stopifnot(all(TCF7_regul %>% dplyr::select(mor) %>% unlist() == 1))
# All targets belong to class C
TCF7_regul %>% dplyr::select(confidence) %>% unlist() %>% table()
# Get targets
TCF7_targets <- TCF7_regul %>% dplyr::select(target) %>% unlist()

# Pre-ranked GSEA on individual genes from 01_DGE
apply(KO_eBay.genes$t, 2, function(rnk) {
		fgsea(list(TCF7_targets=TCF7_targets),
		      rnk)[1,"leadingEdge"] %>% unlist()
	})
```
 
The results of downstream TFs of WNT canonical pathway are not clear.

### WNT Planar Cell Polarity pathway
> ref: Schambony and Wedlich 2007

It seems that JUN is not consistently dysregulated upon NKD2 perturbation because the both KO
do not agree, but ATF2 does.
```{r}
# JUN
dat[dat$ID=="JUN",c(1,3,4,5,7,8)]
# ATF2
dat[dat$ID=="ATF2",c(1,3,4,5,7,8)]
```

## Conclusions
> In agreement with PROGENy, it seems that JAK-STAT transcription factors are dysregulated towards
NK2 functional impact. In addition, it seems that JUN and ATF2, which are downstream TFs from WNT
planar cell polarity pathway, might be dysregulated following the same trend. But JUN is not consistent.


## Save data
```{r}
saveRDS(KO_TF, file=paste0(DATADIR,"/KO_v.rds"))
saveRDS(OE_TF, file=paste0(DATADIR,"/OE_v.rds"))
saveRDS(KO_eBay, file=paste0(DATADIR,"/KO_eBay.rds"))
saveRDS(OE_eBay, file=paste0(DATADIR,"/OE_eBay.rds"))
```

## Session info

```{r}
sessionInfo()

{                                                                                                                                                                                                           
sink(file=paste0(OUTDIR,"/sessionInfo.txt"))
print(sessionInfo())
sink()
}
```
